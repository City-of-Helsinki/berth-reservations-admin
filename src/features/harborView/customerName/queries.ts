import gql from 'graphql-tag';

export const CUSTOMER_NAME_QUERY = gql`
  query CUSTOMER_NAME($customerId: ID!) {
    profile(id: $customerId, serviceType: BERTH) {
      id
      firstName
      lastName
    }
  }
`;
