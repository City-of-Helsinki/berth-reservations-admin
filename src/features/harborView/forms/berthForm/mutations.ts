import gql from 'graphql-tag';

export const UPDATE_BERTH_MUTATION = gql`
  mutation UPDATE_BERTH($input: UpdateBerthMutationInput!) {
    updateBerth(input: $input) {
      clientMutationId
    }
  }
`;

export const DELETE_BERTH_MUTATION = gql`
  mutation DELETE_BERTH($input: DeleteBerthMutationInput!) {
    deleteBerth(input: $input) {
      clientMutationId
    }
  }
`;

export const CREATE_BERTH_MUTATION = gql`
  mutation CREATE_BERTH($input: CreateBerthMutationInput!) {
    createBerth(input: $input) {
      clientMutationId
    }
  }
`;
