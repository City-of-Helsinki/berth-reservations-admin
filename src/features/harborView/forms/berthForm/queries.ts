import gql from 'graphql-tag';

export const INDIVIDUAL_BERTH_QUERY = gql`
  query INDIVIDUAL_BERTH($id: ID!) {
    berth(id: $id) {
      number
      comment
      isActive
      pier {
        id
        properties {
          identifier
        }
      }
      mooringType
      width
      length
      depth
    }
  }
`;
