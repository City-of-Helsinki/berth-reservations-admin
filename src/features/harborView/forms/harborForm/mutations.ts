import { gql } from 'apollo-boost';

export const UPDATE_HARBOR_MUTATION = gql`
  mutation UPDATE_HARBOR($input: UpdateHarborMutationInput!) {
    updateHarbor(input: $input) {
      clientMutationId
    }
  }
`;
