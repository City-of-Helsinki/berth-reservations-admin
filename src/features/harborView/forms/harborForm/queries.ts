import { gql } from 'apollo-boost';

export const HARBOR_FORM_QUERY = gql`
  query HARBOR_FORM($id: ID!) {
    harbor(id: $id) {
      id
      properties {
        name
        streetAddress
        zipCode
        municipality
        wwwUrl
        imageFile
        maps {
          id
          url
        }
      }
    }
  }
`;
