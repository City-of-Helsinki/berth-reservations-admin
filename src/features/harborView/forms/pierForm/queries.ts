import gql from 'graphql-tag';

export const BOAT_TYPES_QUERY = gql`
  query BOAT_TYPES {
    boatTypes {
      id
      name
    }
  }
`;

export const PIER_AND_BOAT_TYPES_QUERY = gql`
  query PIER_AND_BOAT_TYPES($id: ID!) {
    pier(id: $id) {
      properties {
        identifier
        priceTier
        suitableBoatTypes {
          id
        }
        mooring
        lighting
        electricity
        gate
        wasteCollection
        water
        personalElectricity
      }
    }
    boatTypes {
      id
      name
    }
  }
`;
