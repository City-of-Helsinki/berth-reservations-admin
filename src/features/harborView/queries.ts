import gql from 'graphql-tag';

export const INDIVIDUAL_HARBOR_QUERY = gql`
  query INDIVIDUAL_HARBOR($harborId: ID!) {
    harbor(id: $harborId) {
      id
      properties {
        name
        numberOfPlaces
        numberOfFreePlaces
        streetAddress
        zipCode
        municipality
        wwwUrl
        imageFile
        maps {
          id
          url
        }
        servicemapId
        maxWidth
      }
    }
    piers(harbor: $harborId) {
      edges {
        node {
          id
          properties {
            identifier
            priceTier
            electricity
            wasteCollection
            water
            lighting
            gate
            suitableBoatTypes {
              name
            }
          }
        }
      }
    }
    berths(harbor: $harborId) {
      count
      edges {
        node {
          id
          isActive
          number
          width
          length
          depth
          mooringType
          comment
          leases {
            edges {
              node {
                customer {
                  id
                }
                status
                startDate
                endDate
                isActive
              }
            }
          }
          pier {
            id
            properties {
              identifier
            }
          }
        }
      }
    }
  }
`;
