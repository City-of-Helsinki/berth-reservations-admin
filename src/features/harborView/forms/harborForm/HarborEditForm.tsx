import React from 'react';
import { TextInput } from 'hds-react';
import { useTranslation } from 'react-i18next';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import { TFunction } from 'i18next';
import { useMutation, useQuery } from '@apollo/react-hooks';

import styles from './harborEditForm.module.scss';
import Section from '../../../../common/section/Section';
import Text from '../../../../common/text/Text';
import { HARBOR_FORM_QUERY } from './queries';
import LoadingSpinner from '../../../../common/spinner/LoadingSpinner';
import { FormProps, Harbor } from '../types';
import { getHarbor, mapValuesToMutation } from './utils/utils';
import { UPDATE_HARBOR_MUTATION } from './mutations';
import { UPDATE_HARBOR, UPDATE_HARBORVariables as UPDATE_HARBOR_VARS } from './__generated__/UPDATE_HARBOR';
import { HARBOR_FORM } from './__generated__/HARBOR_FORM';
import Button from '../../../../common/button/Button';

export interface Props extends FormProps<Harbor> {
  harborId: string;
}

const getValidationSchema = (t: TFunction) =>
  Yup.object<Harbor>().shape({
    name: Yup.string().required(t('forms.common.errors.required')),
    streetAddress: Yup.string().required(t('forms.common.errors.required')),
    zipCode: Yup.string().required(t('forms.common.errors.required')),
    municipality: Yup.string().required(t('forms.common.errors.required')),
    wwwUrl: Yup.string().required(t('forms.common.errors.required')),
  });

const HarborEditForm = ({ harborId, onCancel, onSubmit, refetchQueries }: Props) => {
  const { loading, error, data } = useQuery<HARBOR_FORM>(HARBOR_FORM_QUERY, {
    variables: { id: harborId },
  });
  const [updateHarbor, { loading: isSubmitting, error: updateError }] = useMutation<UPDATE_HARBOR, UPDATE_HARBOR_VARS>(
    UPDATE_HARBOR_MUTATION,
    {
      refetchQueries: [...(refetchQueries ?? []), { query: HARBOR_FORM_QUERY, variables: { id: harborId } }],
    }
  );

  const { t } = useTranslation();
  const validationSchema = getValidationSchema(t);

  if (loading) return <LoadingSpinner isLoading={loading} />;
  if (error || updateError) return <div>{t('forms.common.error')}</div>;

  const initial: Harbor = getHarbor(data) ?? {
    name: '',
    streetAddress: '',
    zipCode: '',
    municipality: 'Helsinki',
    wwwUrl: '',
  };

  const handleSubmit = (values: Harbor) =>
    updateHarbor({
      variables: {
        input: mapValuesToMutation(harborId, values),
      },
    }).then(() => onSubmit?.(values));

  return (
    <Formik initialValues={initial} onSubmit={handleSubmit} validationSchema={validationSchema}>
      {({ errors, setFieldValue, values }) => (
        <Form className={styles.harborEditForm}>
          <Text as="h4" color="brand">
            {t('forms.harbor.title').toUpperCase()}
          </Text>

          <Section>
            <Field
              required={true}
              as={TextInput}
              id="name"
              name="name"
              label={t('forms.harbor.name')}
              invalid={!!errors.name}
              helperText={errors.name}
            />
          </Section>

          <Section>
            <Field
              required={true}
              as={TextInput}
              id="streetAddress"
              name="streetAddress"
              label={t('forms.harbor.streetAddress')}
              invalid={!!errors.streetAddress}
              helperText={errors.streetAddress}
            />
            <Field
              required={true}
              as={TextInput}
              id="zipCode"
              name="zipCode"
              label={t('forms.harbor.zipCode')}
              invalid={!!errors.zipCode}
              helperText={errors.zipCode}
            />
            <Field
              required={true}
              as={TextInput}
              id="municipality"
              name="municipality"
              label={t('forms.harbor.municipality')}
              invalid={!!errors.municipality}
              helperText={errors.municipality}
            />
          </Section>

          <Section>
            <Field
              required={true}
              as={TextInput}
              id="wwwUrl"
              name="wwwUrl"
              label={t('forms.harbor.wwwUrl')}
              invalid={!!errors.wwwUrl}
              helperText={errors.wwwUrl}
            />
          </Section>

          <div className={styles.formActionButtons}>
            <Button variant="secondary" onClick={onCancel}>
              {t('common.cancel')}
            </Button>
            <Button type="submit" disabled={isSubmitting}>
              {t('common.save')}
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default HarborEditForm;
