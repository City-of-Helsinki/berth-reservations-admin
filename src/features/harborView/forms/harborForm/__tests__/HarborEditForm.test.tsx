import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import waitForExpect from 'wait-for-expect';
import { MockedProvider } from '@apollo/react-testing';

import LoadingSpinner from '../../../../../common/spinner/LoadingSpinner';
import HarborEditForm from '../HarborEditForm';
import { HARBOR_FORM_QUERY } from '../queries';
import { UPDATE_HARBOR_MUTATION } from '../mutations';

const queryMock = {
  request: { query: HARBOR_FORM_QUERY, variables: { id: 'a' } },
  result: {
    data: {
      harbor: {
        __typename: 'HarborNode',
        id: 'a',
        properties: {
          __typename: 'HarborProperties',
          name: 'Pajalahden venesatama (Meripuistotie) / Venesatama',
          streetAddress: 'Meripuistotie 1a',
          zipCode: '00210',
          municipality: 'Helsinki',
          wwwUrl: 'https://hel.fi',
          imageFile: 'https://hel.fi',
          maps: [
            {
              __typename: 'HarborMapType',
              id: 'aaa',
              url: 'https://hel.ninja',
            },
          ],
        },
      },
    },
  },
};

describe('HarborEditForm', () => {
  const waitForContent = async (wrapper: ReactWrapper) => {
    await act(async () => {
      await waitForExpect(() => {
        wrapper.update();
        expect(wrapper.contains(<LoadingSpinner />)).toBeFalsy();
      });
    });
  };

  it('initially renders loading spinner', () => {
    const wrapper = mount(
      <MockedProvider mocks={[queryMock]}>
        <HarborEditForm onCancel={jest.fn()} onSubmit={jest.fn()} harborId="a" />
      </MockedProvider>
    );
    expect(wrapper.contains(<LoadingSpinner />)).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders content after loading', async () => {
    const wrapper = mount(
      <MockedProvider mocks={[queryMock]}>
        <HarborEditForm onCancel={jest.fn()} onSubmit={jest.fn()} harborId="a" />
      </MockedProvider>
    );
    await waitForContent(wrapper);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('calls update mutation on save', async () => {
    let updateMockCalled = false;
    const onUpdateMock = jest.fn();
    const updateMock = {
      request: {
        query: UPDATE_HARBOR_MUTATION,
        variables: {
          input: {
            id: 'a',
            name: 'Pajalahden venesatama (Meripuistotie) / Venesatama',
            streetAddress: 'Meripuistotie 1a',
            zipCode: '00210',
            municipalityId: 'helsinki',
            wwwUrl: 'https://hel.fi',
          },
        },
      },
      result: () => {
        updateMockCalled = true;
        return {
          data: { updateHarbor: { clientMutationId: '-', __typename: 'ID' } },
        };
      },
    };
    const wrapper = mount(
      // We need queryMock twice here, because MockedProvider requires an
      // instance for each query made and the original query is refetched after updates.
      <MockedProvider mocks={[queryMock, queryMock, updateMock]}>
        <HarborEditForm onCancel={jest.fn()} onSubmit={onUpdateMock} harborId="a" />
      </MockedProvider>
    );
    await waitForContent(wrapper);
    await act(async () => {
      wrapper.find('Form').simulate('submit');
      await waitForExpect(() => {
        wrapper.update();
        expect(onUpdateMock).toBeCalledTimes(1);
        expect(updateMockCalled).toBe(true);
      });
    });
  });
});
