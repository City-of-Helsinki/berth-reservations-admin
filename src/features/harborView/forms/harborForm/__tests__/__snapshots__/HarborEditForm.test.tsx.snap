// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HarborEditForm initially renders loading spinner 1`] = `"<div class=\\"spinnerWrapper\\"><div class=\\"spinner\\"></div></div>"`;

exports[`HarborEditForm renders content after loading 1`] = `"<form action=\\"#\\" class=\\"harborEditForm\\"><h4 class=\\"text brand h4\\">SATAMA</h4><article class=\\"section\\"><section class=\\"body\\"><div class=\\"TextInput-module_root__2CMNr text-input_hds-text-input__2LODq\\"><label for=\\"name\\" class=\\"FieldLabel-module_label__2tl-4 text-input_hds-text-input__label__pDw_W \\">Nimi<span aria-hidden=\\"true\\" class=\\"RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs\\">*</span></label><div class=\\"TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG\\"><input class=\\"TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C\\" id=\\"name\\" required=\\"\\" type=\\"text\\" name=\\"name\\" value=\\"Pajalahden venesatama (Meripuistotie) / Venesatama\\"></div></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"TextInput-module_root__2CMNr text-input_hds-text-input__2LODq\\"><label for=\\"streetAddress\\" class=\\"FieldLabel-module_label__2tl-4 text-input_hds-text-input__label__pDw_W \\">Osoite<span aria-hidden=\\"true\\" class=\\"RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs\\">*</span></label><div class=\\"TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG\\"><input class=\\"TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C\\" id=\\"streetAddress\\" required=\\"\\" type=\\"text\\" name=\\"streetAddress\\" value=\\"Meripuistotie 1a\\"></div></div><div class=\\"TextInput-module_root__2CMNr text-input_hds-text-input__2LODq\\"><label for=\\"zipCode\\" class=\\"FieldLabel-module_label__2tl-4 text-input_hds-text-input__label__pDw_W \\">Postinumero<span aria-hidden=\\"true\\" class=\\"RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs\\">*</span></label><div class=\\"TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG\\"><input class=\\"TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C\\" id=\\"zipCode\\" required=\\"\\" type=\\"text\\" name=\\"zipCode\\" value=\\"00210\\"></div></div><div class=\\"TextInput-module_root__2CMNr text-input_hds-text-input__2LODq\\"><label for=\\"municipality\\" class=\\"FieldLabel-module_label__2tl-4 text-input_hds-text-input__label__pDw_W \\">Postitoimipaikka<span aria-hidden=\\"true\\" class=\\"RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs\\">*</span></label><div class=\\"TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG\\"><input class=\\"TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C\\" id=\\"municipality\\" required=\\"\\" type=\\"text\\" name=\\"municipality\\" value=\\"Helsinki\\"></div></div></section></article><article class=\\"section\\"><section class=\\"body\\"><div class=\\"TextInput-module_root__2CMNr text-input_hds-text-input__2LODq\\"><label for=\\"wwwUrl\\" class=\\"FieldLabel-module_label__2tl-4 text-input_hds-text-input__label__pDw_W \\">WWW-osoite<span aria-hidden=\\"true\\" class=\\"RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs\\">*</span></label><div class=\\"TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG\\"><input class=\\"TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C\\" id=\\"wwwUrl\\" required=\\"\\" type=\\"text\\" name=\\"wwwUrl\\" value=\\"https://hel.fi\\"></div></div></section></article><div class=\\"formActionButtons\\"><button type=\\"button\\" class=\\"Button-module_button__3wgee button_hds-button__2A0je Button-module_secondary__399Vd button_hds-button--secondary__1NOWS Button-module_theme-coat__2xA5D button_hds-button--theme-coat__12cO0\\"><span class=\\"Button-module_label__2zMLC button_hds-button__label__2EQa-\\">Peruuta</span></button><button type=\\"submit\\" class=\\"Button-module_button__3wgee button_hds-button__2A0je Button-module_primary__3bo_T button_hds-button--primary__2NVvO Button-module_theme-coat__2xA5D button_hds-button--theme-coat__12cO0\\"><span class=\\"Button-module_label__2zMLC button_hds-button__label__2EQa-\\">Tallenna</span></button></div></form>"`;
